File: android/src/org/jetbrains/android/intentions/AddActivityToManifest.java
/*
 * Copyright (C) 2016 The Android Open Source Project
 *
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 *      http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */
package org.jetbrains.android.intentions;

import com.android.SdkConstants;
import com.intellij.psi.PsiClass;
import com.intellij.psi.util.InheritanceUtil;
import org.jetbrains.android.dom.manifest.Activity;
import org.jetbrains.android.dom.manifest.Application;
import org.jetbrains.android.dom.manifest.Manifest;
import org.jetbrains.annotations.Nls;
import org.jetbrains.annotations.NotNull;

/**
 * For inheritor of android.app.Activity which isn't registered in the AndroidManifest.xml file,
 * add an intention action to register it.
 */
public class AddActivityToManifest extends AbstractRegisterComponentAction {
  @Nls
  @NotNull
  @Override
  public String getText() {
    return "Add activity to manifest";
  }

  @Override
  boolean isAvailable(@NotNull PsiClass psiClass, @NotNull Manifest manifest) {
    // Check whether the class is a subclass of Activity
    if (!(InheritanceUtil.isInheritor(psiClass, true, SdkConstants.CLASS_ACTIVITY))) {
      return false;
    }

    // Check whether this Activity is already registered
    return !isRegisteredActivity(psiClass, manifest.getApplication());
  }

  @Override
  void invoke(@NotNull PsiClass psiClass, @NotNull Manifest manifest) {
    final Application application = manifest.getApplication();
    application.addActivity().getActivityClass().setValue(psiClass);
  }

  private static boolean isRegisteredActivity(PsiClass psiClass, Application application) {
    for (Activity activity : application.getActivities()) {
      if (psiClass.isEquivalentTo(activity.getActivityClass().getValue())) {
        // Found existing activity with the same class
        return true;
      }
    }
    return false;
  }
}

File: android-kotlin/idea-android/src/org/jetbrains/kotlin/android/intention/AddActivityToManifest.kt
/*
 * Copyright 2010-2017 JetBrains s.r.o.
 *
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 * http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */

package org.jetbrains.kotlin.android.intention

import com.android.SdkConstants
import org.jetbrains.android.dom.manifest.Manifest
import com.android.tools.idea.kotlin.isSubclassOf
import org.jetbrains.kotlin.analysis.api.KtAllowAnalysisOnEdt
import org.jetbrains.kotlin.analysis.api.analyze
import org.jetbrains.kotlin.analysis.api.lifetime.allowAnalysisOnEdt
import org.jetbrains.kotlin.android.isSubclassOf
import org.jetbrains.kotlin.descriptors.ClassDescriptor
import org.jetbrains.kotlin.idea.search.usagesSearch.descriptor
import org.jetbrains.kotlin.psi.KtClass
import org.jetbrains.kotlin.asJava.toLightClass
import org.jetbrains.kotlin.idea.base.plugin.isK2Plugin


class AddActivityToManifest : AbstractRegisterComponentAction("Add activity to manifest") {
    override fun isApplicableTo(element: KtClass, manifest: Manifest): Boolean =
            element.isSubclassOfActivity() && !element.isRegisteredActivity(manifest)

    override fun applyTo(element: KtClass, manifest: Manifest) {
        val psiClass = element.toLightClass() ?: return
        manifest.application.addActivity().activityClass.value = psiClass
    }

    private fun KtClass.isRegisteredActivity(manifest: Manifest) = manifest.application.activities.any {
        it.activityClass.value?.qualifiedName == fqName?.asString()
    }

    @OptIn(KtAllowAnalysisOnEdt::class)
    private fun KtClass.isSubclassOfActivity() = if (isK2Plugin()) {
        allowAnalysisOnEdt {
            analyze(this@isSubclassOfActivity) {
                isSubclassOf(this@isSubclassOfActivity, SdkConstants.CLASS_ACTIVITY, strict = true)
            }
        }
    }
    else {
        (descriptor as? ClassDescriptor)?.defaultType?.isSubclassOf(SdkConstants.CLASS_ACTIVITY, strict = true) ?: false
    }
}
